File: YACS Python Solution/2020.10 YACS 丙组乙组甲组/YACS 2020.10 乙组 #4.cpp
#include<bits/stdc++.h>
using namespace std;
int n,m,a[100002],ll,rr,mid;

bool check(int minh){
    int value=0,j=0;
    for (int i = 0; i < n; i++)
    {
        while(j<m && value<minh) {
            value = value+a[j];
            j++;
        }
        if (value<minh) return false;
        value = value/2;
    }
    return true;
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin>>n>>m;
    for (int i = 0; i < m; i++) {
        cin>>a[i];
        rr = rr+a[i];
    }
    while(rr>ll+1){
        mid = (rr+ll)/2;
        if(check(mid)) ll=mid;
        else rr=mid;
    }
    cout<<ll<<endl;
}

File: YACS Python Solution/2020.10 YACS 丙组乙组/YACS 2020.10 乙组 #4.py
'''
标准输出: 1991
6 5
147 1969 933 10095 1557

标准输出: 4843
5 6
147 1969 933 10095 1557 10308

标准输出: 7913
7 7
199 10104 11847 1892 2750 7600 9804

'''
# 二分搜索
def check(minh):
    value,j = 0,0
    for _ in range(n):
        while j<m and value<minh:
            value+=a[j]
            j+=1
        if value<minh: return False
        value//=2
    return True
    
n,m = map(int, input().split())
a = list(map(int, input().split()))

ll,rr = 0,sum(a)//2+1
while rr>ll+1:
    mid = (rr+ll)//2
    if check(mid): ll = mid
    else: rr = mid
print(ll)